src/common/io/tls/session.c

```diff
@@ -0,0 +1,304 @@
+        sckSessionFree(this->socketSession);
+        this->socketSession = NULL;
+
+        // Free the TLS session
+        memContextCallbackClear(this->memContext);
+        tlsSessionFreeResource(this);
+        this->session = NULL;
+    }
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/***********************************************************************************************************************************
+Report TLS errors.  Returns true if the command should continue and false if it should exit.
+***********************************************************************************************************************************/
+static bool
+tlsSessionError(TlsSession *this, int code)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(TLS_SESSION, this);
+        FUNCTION_LOG_PARAM(INT, code);
+    FUNCTION_LOG_END();
+
+    bool result = false;
+
+    switch (code)
+    {
+        // The connection was closed
+        case SSL_ERROR_ZERO_RETURN:
+        {
+            tlsSessionClose(this);
+            break;
+        }
+
+        // Try the read/write again
+        case SSL_ERROR_WANT_READ:
+        case SSL_ERROR_WANT_WRITE:
+        {
+            result = true;
+            break;
+        }
+
+        // A syscall failed (this usually indicates eof)
+        case SSL_ERROR_SYSCALL:
+        {
+            // Get the error before closing so it is not cleared
+            int errNo = errno;
+            tlsSessionClose(this);
+
+            // Throw the sys error if there is one
+            THROW_ON_SYS_ERROR(errNo, KernelError, "tls failed syscall");
+
+            break;
+        }
+
+        // Some other tls error that cannot be handled
+        default:
+            THROW_FMT(ServiceError, "tls error [%d]", code);
+    }
+
+    FUNCTION_LOG_RETURN(BOOL, result);
+}
+
+/***********************************************************************************************************************************
+Read from the TLS session
+***********************************************************************************************************************************/
+static size_t
+tlsSessionRead(THIS_VOID, Buffer *buffer, bool block)
+{
+    THIS(TlsSession);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(TLS_SESSION, this);
+        FUNCTION_LOG_PARAM(BUFFER, buffer);
+        FUNCTION_LOG_PARAM(BOOL, block);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->session != NULL);
+    ASSERT(buffer != NULL);
+    ASSERT(!bufFull(buffer));
+
+    ssize_t result = 0;
+
+    // If blocking read keep reading until buffer is full
+    do
+    {
+        // If no tls data pending then check the socket
+        if (!SSL_pending(this->session))
+            sckSessionReadWait(this->socketSession);
+
+        // Read and handle errors
+        size_t expectedBytes = bufRemains(buffer);
+        result = SSL_read(this->session, bufRemainsPtr(buffer), (int)expectedBytes);
```

Comment:
Good question! I imagine this is some cruft left over from a prior refactor. It's not really part of this refactor so I'll commit it separately.

Reply:
Committed at b5347070af5e and b7d8d61526.

We'll need to make sure this does not creep back in since the next commit you are reviewing copies a lot of code. Please either comment on your next review or we can leave this comment open until we know it is fixed.

Reply:
I made sure this change got into the final "Split session functionality of TlsClient" commit. Still in CI so I'll post the commit once it is done.

Reply:
I confirmed in master, so no need for the commit number. I'm resolving this conversation.